File: src/swift/4_BehavioralPatterns/2_Command/Exercise.swift
import Foundation
import XCTest

class Command
{
  enum Action
  {
    case deposit
    case withdraw
  }

  var action: Action
  var amount: Int
  var success = false

  init(_ action: Action, _ amount: Int)
  {
    self.action = action
    self.amount = amount
  }
}

class Account
{
  var balance = 0

  func process(_ c: Command)
  {
    switch c.action
    {
      case .deposit:
        balance += c.amount
        c.success = true
      case .withdraw:
        c.success = (balance >= c.amount)
        if (c.success) { balance -= c.amount }
    }
  }
}

class UMBaseTestCase : XCTestCase {}

//@testable import Test

class Evaluate: UMBaseTestCase
{
  func simpleTest()
  {
    let a = Account()
    
    let deposit = Command(.deposit, 100)
    a.process(deposit)

    XCTAssertEqual(100, a.balance, "Expected the balance to rise to 100")
    XCTAssert(deposit.success, "A deposit should always succeed")

    let withdraw = Command(.withdraw, 50)
    a.process(withdraw)

    XCTAssertEqual(50, a.balance)
    XCTAssert(withdraw.success, "Withdrawal of 50 should have succeeded")

    let withdraw2 = Command(.withdraw, 100)
    a.process(withdraw2)

    XCTAssertEqual(50, a.balance, "After a failed withdrawal, balance should have stayed at exactly 50")
    XCTAssertFalse(withdraw2.success, "Attempted withdrawal should have failed")
  }
}

extension Evaluate
{
  static var allTests : [(String, (Evaluate) -> () throws -> Void)]
  {
    return [
      ("simpleTest", simpleTest)
    ]
  }
}

func main()
{
  XCTMain([testCase(Evaluate.allTests)])
}

main()

File: src/python/4_BehavioralPatterns/2_Command/exercise.py
import unittest
from enum import Enum


class Command:
    class Action(Enum):
        DEPOSIT = 0
        WITHDRAW = 1

    def __init__(self, action, amount):
        self.action = action
        self.amount = amount
        self.success = False


class Account:
    def __init__(self, balance=0):
        self.balance = balance

    def process(self, command):
        if command.action == Command.Action.DEPOSIT:
            self.balance += command.amount
            command.success = True
        elif command.action == Command.Action.WITHDRAW:
            command.success = self.balance >= command.amount
            if command.success:
                self.balance -= command.amount


class Evaluate(unittest.TestCase):
    def test(self):
        a = Account()

        cmd = Command(Command.Action.DEPOSIT, 100)
        a.process(cmd)

        self.assertEqual(100, a.balance)
        self.assertTrue(cmd.success)

        cmd = Command(Command.Action.WITHDRAW, 50)
        a.process(cmd)

        self.assertEqual(50, a.balance)
        self.assertTrue(cmd.success)

        cmd.amount = 150
        a.process(cmd)

        self.assertEqual(50, a.balance)
        self.assertFalse(cmd.success)
